c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,433 @@
+    }
+
+    void operator++() {
+      ++value;
+    }
+
+    bool operator==(iterator const& other) const = default;
+    bool operator!=(iterator const& other) const = default;
+
+    char32_t value;
+  };
+
+  iterator begin() const {
+    return {front_};
+  }
+
+  iterator end() const {
+    return {back_ + 1};
+  }
+
+ private:
+  char32_t front_;
+  char32_t back_;
+};
+
+constexpr CodePointRange convex_hull(CodePointRange const& r1,
+                                     CodePointRange const& r2) {
+  return CodePointRange(std::min(r1.front(), r2.front()),
+                        std::max(r1.back(), r2.back()));
+}
+
+class CodePointSet {
+ public:
+  // Empty set.
+  CodePointSet() = default;
+
+  bool contains(char32_t c) const {
+    auto const it = std::partition_point(
+        ranges_.begin(), ranges_.end(), [c](CodePointRange const& range) {
+          return range.back() < c;
+        });
+    return it != ranges_.end() && it->contains(c);
+  }
+
+  void Include(CodePointRange range) {
+    auto it = std::partition_point(
+        ranges_.begin(), ranges_.end(), [&range](CodePointRange const& r) {
+          return r.back() < range.front();
+        });
+    if (it == ranges_.end() || it->intersection(range).empty()) {
+      it = ranges_.insert(it, range);
+      if ((it + 1) != ranges_.end() && (it + 1)->front() == it->back() + 1) {
+        *it = convex_hull(*it, *(it + 1));
+        ranges_.erase(it + 1);
+      }
+      if (it != ranges_.begin() && (it - 1)->back() + 1 == it->front()) {
+        *(it - 1) = convex_hull(*(it - 1), *it);
+        ranges_.erase(it);
+      }
+    } else {
+      *it = convex_hull(*it, range);
+    }
+  }
+
+  CodePointSet Include(CodePointSet const& other) {
+    for (auto const& range : other.ranges_) {
+      Include(range);
+    }
+  }
+
+ private:
+  // Non-overlapping ranges in ascending order.
+  std::vector<CodePointRange> ranges_;
+};
+
+constexpr std::string_view strip_comment(std::string_view const line) {
+  return line.substr(0, line.find_first_of('#'));
+}
+
+template<std::size_t n>
+constexpr std::array<std::string_view, n> fields(std::string_view const line) {
+  std::array<std::string_view, n> result;
+  auto split = std::views::split(line, ';');
+  auto it = split.begin();
+  for (auto& field : result) {
+    CHECK(it != split.end()) << line;
+    std::string_view field_with_spaces(*it++);
+    auto const begin = std::min(field_with_spaces.find_first_not_of(" "),
+                                field_with_spaces.size());
+    auto const end = field_with_spaces.find_last_not_of(" ") + 1;
+    field = field_with_spaces.substr(begin, end - begin);
+  }
+  return result;
+}
+
+constexpr char32_t parse_hex_codepoint(std::string_view hex) {
+  std::uint32_t c;
+  std::from_chars(hex.data(), hex.data() + hex.size(), c, 16);
+  return static_cast<char32_t>(c);
+}
+
+constexpr CodePointRange parse_hex_codepoint_range(std::string_view hex) {
+  auto const front_end = hex.find("..");
```

Comment:
confusing... I would call it "pos"

Reply:
Yeah this is not a good name here (it is not even the end of the text representation of the variable called `front` below!).